### ISSUE TITLE
نصب اولیه

### ISSUE THREAD

**Issue by FatemehPasban** [FatemehPasban/bert-nmt#1]

**Comment by FatemehPasban** [comment]
برای نصب ابتدا پایتورچ نسخه ۱.۱۲.۰  نصب شد سپس fairseq  ۰.۶.۲ هردو با استفاده از pip .

برای ترین جدید مدل باید ۳ گام انجام گیرد :
۱. ابتدا فایل داده ها تولید میشه با اجرای  یکی از فایل های اجرایی prepare...sh که از دیتاست های اماده استفاده میکنه برای اینکه بدونیم داده های ما به چه شکا باید در بیان یکی از این فایل ها رو تولید کردیم تا بخ عنوان نمونه بشه ازش استفاده کرد.
چیزی که من در نتیجه ی اجرای این کد دیدم یک فایل بدون فرمت هست  که دیتایی برحسب جمله با تفکیک اسپیس هست. یعنی هر خط یک داده از مجموعه دیتاهای دیتاست محسوب میشه.(پس برای اینکه از داده های خودمون بخوایم استفاده کنیم باید داده هارو به همین فرم تبدیل کنیم یعنی فایلی که پسوندش زبان داده هاست. مثلا فایل train.fa   train.en  فایل داده های آموزش هستند که داخل هر فایل چندین جمله داریم  و هر جمله با رفتن به خط بعد تفکیک شده)

۲. در گام ۲ قراره با استفاده از ابزار  mosesdecoder  (part of moses) داده های گام ۱ برای ورودی مدل برت آماده بشن (اسم فایلی که این کار رو انجام میده detokenizer.perl هست که به زبان perl نوشته شده) فقط نکته ای که هست برای اینکار باید فایل اجرایی makedataforbert.sh  در کنار  detokenizer.perl  و فایل های تولید شده در گام ۱ باهم قرار بگیرند. و خروجی فایل هایی با نام های : 

test.bert.en , test.bert.de , train.bert.en,  train.bert.de , valid.bert.en, valid.bert.de  خواهد بود.

**Comment by FatemehPasban** [comment]
بعد از اینکه داده ها تقریبا آماده شدن یک گام نهایی باید روی آنها انجام بشه، و فایل هایی برای آموزش ،... آماده بشه این دفعه این فایل ها در مسیری که به عنوان destdir در دستور زیر مشخص کرده ایم قرار خواهند گرفت: 

$python preprocess.py --source-lang en --target-lang de   --trainpref $TEXT/train --validpref $TEXT/valid --testpref $TEXT/test   --destdir processedData  --joined-dictionary --bert-model-name bert-base-uncased

******************برای اجرای درست دستور بالا. باید  ۱۲ فایلی که تولید شده بود در مسیر root سیستم قرار بدیم.(مسیری که در فایل ها برای read کردن فایل قرار داده شده در مسیر root است.)

**Comment by FatemehPasban** [comment]
******************برای اجرای درست دستور بالا. باید  ۱۲ فایلی که تولید شده بود در مسیر root سیستم قرار بدیم.(مسیری که در فایل ها برای read کردن فایل قرار داده شده در مسیر root است.)

**Comment by FatemehPasban** [comment]
۳. در گام بعد باید فایل trajn  رو توسط یک فایل اجرایی دیگر اجرا کنیم.
#!/usr/bin/env bash
nvidia-smi

cd /home/fatemeh/PycharmProjects/bert-nmt
python3 -c "import torch; print(torch.__version__)"

src=en
tgt=de
bedropout=0.5
ARCH=transformer_s2_iwslt_de_en
DATAPATH=/home/fatemeh/PycharmProjects/bert-nmt/processedData
SAVEDIR=checkpoints/iwed_${src}_${tgt}_${bedropout}
mkdir -p $SAVEDIR
if [ ! -f $SAVEDIR/checkpoint_nmt.pt ]
then
    cp /your_pretrained_nmt_model $SAVEDIR/checkpoint_nmt.pt
fi
if [ ! -f "$SAVEDIR/checkpoint_last.pt" ]
then
warmup="--warmup-from-nmt --reset-lr-scheduler"
else
warmup=""
fi

python train.py $DATAPATH -a $ARCH --optimizer adam --lr 0.0005 -s $src -t $tgt --label-smoothing 0.1 \
--dropout 0.3 --max-tokens 4000 --min-lr '1e-09' --lr-scheduler inverse_sqrt --weight-decay 0.0001 \
--criterion label_smoothed_cross_entropy --max-update 150000 --warmup-updates 4000 --warmup-init-lr '1e-07' \
--adam-betas '(0.9,0.98)' --save-dir $SAVEDIR --share-all-embeddings $warmup \
--encoder-bert-dropout --encoder-bert-dropout-ratio $bedropout | tee -a $SAVEDIR/training.log
